src/bio/bio_xstream.c

```diff
@@ -1384,13 +1454,89 @@ bio_xsctxt_free(struct bio_xs_context *ctxt)
 	D_FREE(ctxt);
 }
 
+/*
+ * Finalize per-xstream NVMe context.
+ *
+ * \param[IN] ctxt	Per-xstream NVMe context
+ *
+ * \returns		N/A
+ */
+void
+bio_xsctxt_free(struct bio_xs_context *ctxt)
+{
+	int			 rc = 0;
+	enum smd_dev_type	 st;
+	struct bio_xs_blobstore	*bxb;
+
+	/* NVMe context setup was skipped */
+	if (ctxt == NULL)
+		return;
+
+	ctxt->bxc_ready = 0;
+	for (st = SMD_DEV_TYPE_DATA; st < SMD_DEV_TYPE_MAX; st++) {
+		bxb = ctxt->bxc_xs_blobstores[st];
+		if (bxb == NULL)
+			continue;
+
+		if (bxb->bxb_io_channel != NULL) {
+			spdk_bs_free_io_channel(bxb->bxb_io_channel);
+			bxb->bxb_io_channel = NULL;
+		}
+		if (bxb->bxb_blobstore != NULL) {
+			bio_blobstore_free(bxb, ctxt);
+			bxb->bxb_blobstore = NULL;
+		}
+		D_FREE(bxb);
+		ctxt->bxc_xs_blobstores[st] = NULL;
+	}
+
+	ABT_mutex_lock(nvme_glb.bd_mutex);
+	if (nvme_glb.bd_xstream_cnt > 0)
+		nvme_glb.bd_xstream_cnt--;
+
+	if (nvme_glb.bd_init_thread != NULL && nvme_glb.bd_xstream_cnt == 0)
+		ABT_cond_broadcast(nvme_glb.bd_barrier);
+	ABT_mutex_unlock(nvme_glb.bd_mutex);
+
+	if (ctxt->bxc_thread != NULL) {
+		D_DEBUG(DB_MGMT, "Finalizing SPDK thread, tgt_id:%d",
+			ctxt->bxc_tgt_id);
+
+		/* Don't drain events if spdk_subsystem_fini() timeout */
+		while (rc == 0 && !spdk_thread_is_idle(ctxt->bxc_thread))
+			spdk_thread_poll(ctxt->bxc_thread, 0, 0);
+
+		D_DEBUG(DB_MGMT, "SPDK thread finalized, tgt_id:%d",
+			ctxt->bxc_tgt_id);
+
+		spdk_thread_exit(ctxt->bxc_thread);
+		ctxt->bxc_thread = NULL;
+	}
+
+	if (ctxt->bxc_dma_buf != NULL) {
+		dma_buffer_destroy(ctxt->bxc_dma_buf);
+		ctxt->bxc_dma_buf = NULL;
+	}
+
+	D_FREE(ctxt);
+}
+
+int
+bio_init_xs_data_blobstore_ctxt(struct bio_xs_context *ctxt, int tgt_id)
```

Comment:
I don't quite see why do we need this.

Reply:
This is used for testing: vos_self_nvme_init, it will try to use sys xs for data testing, however in normal case, we don't allocate datablobstore for sys xs before, or maybe i should add a flag to bio_xsctxt_alloc(), indicate we should always allocate DATA blobstore even it is sys xs.

Reply:
I don't think we need a separate function for this. sys xstream for engine could be differentiated from standalone VOS by target ID.